pkg/compact/compact_test.go

```diff
@@ -174,3 +178,483 @@ func BenchmarkGatherNoCompactionMarkFilter_Filter(b *testing.B) {
+					},
+					Thanos: metadata.Thanos{
+						Version: 1,
+						Labels:  map[string]string{"a": "1"},
+					},
+				},
+				{
+					BlockMeta: tsdb.BlockMeta{
+						ULID:    ulid.MustNew(4, nil),
+						MinTime: int64(8 * time.Hour / time.Millisecond),
+						MaxTime: int64(10 * time.Hour / time.Millisecond),
+					},
+					Thanos: metadata.Thanos{
+						Version: 1,
+						Labels:  map[string]string{"a": "1"},
+					},
+				},
+			},
+			expected: planResult{
+				compactionRuns:   1.0,
+				compactionBlocks: 4.0,
+			},
+		},
+		{
+			// In this test case, the metadata is part of two groups.
+			// The first four blocks are compacted in the first run.
+			testName: "two_groups",
+			input: []*metadata.Meta{
+				{
+					BlockMeta: tsdb.BlockMeta{
+						ULID:    ulid.MustNew(0, nil),
+						MinTime: 0,
+						MaxTime: int64(2 * time.Hour / time.Millisecond),
+					},
+					Thanos: metadata.Thanos{
+						Version: 1,
+						Labels:  map[string]string{"a": "1"},
+					},
+				},
+				{
+					BlockMeta: tsdb.BlockMeta{
+						ULID:    ulid.MustNew(1, nil),
+						MinTime: int64(2 * time.Hour / time.Millisecond),
+						MaxTime: int64(4 * time.Hour / time.Millisecond),
+					},
+					Thanos: metadata.Thanos{
+						Version: 1,
+						Labels:  map[string]string{"a": "1"},
+					},
+				},
+				{
+					BlockMeta: tsdb.BlockMeta{
+						ULID:    ulid.MustNew(2, nil),
+						MinTime: int64(4 * time.Hour / time.Millisecond),
+						MaxTime: int64(6 * time.Hour / time.Millisecond),
+					},
+					Thanos: metadata.Thanos{
+						Version: 1,
+						Labels:  map[string]string{"b": "2"},
+					},
+				},
+				{
+					BlockMeta: tsdb.BlockMeta{
+						ULID:    ulid.MustNew(3, nil),
+						MinTime: int64(6 * time.Hour / time.Millisecond),
+						MaxTime: int64(8 * time.Hour / time.Millisecond),
+					},
+					Thanos: metadata.Thanos{
+						Version: 1,
+						Labels:  map[string]string{"b": "2"},
+					},
+				},
+				{
+					BlockMeta: tsdb.BlockMeta{
+						ULID:    ulid.MustNew(4, nil),
+						MinTime: int64(8 * time.Hour / time.Millisecond),
+						MaxTime: int64(16 * time.Hour / time.Millisecond),
+					},
+					Thanos: metadata.Thanos{
+						Version: 1,
+						Labels:  map[string]string{"b": "2"},
+					},
+				},
+				{
+					BlockMeta: tsdb.BlockMeta{
+						ULID:    ulid.MustNew(6, nil),
+						MinTime: int64(16 * time.Hour / time.Millisecond),
+						MaxTime: int64(18 * time.Hour / time.Millisecond),
+					},
+					Thanos: metadata.Thanos{
+						Version: 1,
+						Labels:  map[string]string{"b": "2"},
+					},
+				},
+				{
+					BlockMeta: tsdb.BlockMeta{
+						ULID:    ulid.MustNew(7, nil),
+						MinTime: int64(18 * time.Hour / time.Millisecond),
+						MaxTime: int64(26 * time.Hour / time.Millisecond),
+					},
+					Thanos: metadata.Thanos{
+						Version: 1,
+						Labels:  map[string]string{"b": "2"},
+					},
+				},
+			},
+			expected: planResult{
+				compactionRuns:   1.0,
+				compactionBlocks: 4.0,
+			},
```

Comment:
Can you also check the result metrics for each group as well?

Reply:
I'm doing that specifically for the two_groups_test in my latest commit.